### ISSUE TITLE
After python autosave, cursor moves to beginning of line

### ISSUE THREAD

**Issue by dacort** [zed-industries/zed#9787]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When autosave is enabled, saving a Python file with an empty method definition results in the cursor moving from an indented position to the beginning of the line, resulting in a frustrating dev experience.

Autosave setting:

```json
  "autosave": {
    "after_delay": {
      "milliseconds": 1000
    }
  }
```

To reproduce:
1. Create a new python file
2. Start to populate with some Python class
```python
class TestClass:
    def some_method(self):
        
```
3. Wait for autosave
4. Observe the cursor moving to the beginning of the line

Note that this happens even if the method definition is in the middle of a file. It appears to be related to the `remove_trailing_whitespace_on_save` setting - if I disable that, then the cursor remains in the same/correct position.

Similar to #7570 and #392 

### Environment

Zed: v0.127.4 (Zed)
OS: macOS 14.3.1
Memory: 36 GiB
Architecture: aarch64

**Comment by JosephTLyons** [comment]
Hey @dacort, can you capture a screen video? I haven't been able to reproduce this on my end.

**Comment by dacort** [comment]
Sure thing!

https://github.com/zed-industries/zed/assets/1512/5423fc5e-48fd-4f46-92ee-a21afcbc9bdb

Note that the file must already exist - in an unsaved buffer, this doesn't happen.

**Comment by JosephTLyons** [comment]
> Sure thing!
> 
>  zed_repro.mov 
> 
> Note that the file must already exist - in an unsaved buffer, this doesn't happen.

Oh ok, I see what you are experiencing now, I was typing before the autosave triggered, so the line was non-empty for me. Yeah this is sort of a tricky situation because that setting is meant to strip the trailing whitespace off all lines. What would be a good fix to this? An option in the autosaver to not run those operations that trim newlines and whitespace?

**Comment by dacort** [comment]
> What would be a good fix to this?

This is a tough one since it's also language-dependent.  I can think of a couple things off-hand:
1. Don't modify the line the cursor is on, since it's likely the user is actively editing that line. A similar thing happens if I start to add a code comment (`# `), pause to think, and then keep typing. But the strip trailing whitespace feature removes my space while I'm thinking.
2. If the user has enabled their own external formatter, should this setting be disabled?
3. As you mentioned, don't run autoformat on autosave

I kind of like option 3 or a variant of it as it seems the most flexible / least confusing. This is a workflow I use in other editors as well - I prefer the delayed autosave, particularly if I'm reloading something in the background, but like that I can manually trigger a format by saving manually. FWIW, this is the setting in VS Code. It doesn't run format on save if autosave with a delay is enabled.

<img width="469" alt="image" src="https://github.com/zed-industries/zed/assets/1512/98328b44-aea9-49f7-946f-b96f49121ecd">

**Comment by iamraphson** [comment]
It is happening in SQL autosave too.

**Comment by JosephTLyons** [comment]
This is now available in Zed [v0.136.0-pre](https://zed.dev/releases/preview/0.136.0).
